### PR TITLE
fix: match a repeated question on word boundaries

### PR DESCRIPTION
**Pull request by aichy126**
## Why this exists as a separate PR

This commit was written for #2917, reviewed there, and replied to in that thread — but it never reached `main`. The branch was updated on GitHub while the commit was only local, so the push that should have carried it was rejected, and #2917 merged from the state before it. `main` today still has the version Devin flagged.

## The problem

`_condensed` removed whitespace rather than reducing it, and the match was plain substring membership. Two consequences, both of which drop a question the learner needs:

- `rate` is contained in `separate`, so a lesson that had only mentioned separating examples would lose the question it meant to ask.
- With spaces removed, words run together, so a match could also straddle two of them.

Either way the learner is left with controls and nothing above them — the outcome this dedupe exists to prevent, and worse than the duplicate it was written to remove.

## The change

- `_condensed` reduces each run of whitespace to a single space instead of dropping it.
- A prompt counts as repeated only where it stands on its own: `_stands_alone` requires the character on either side of a match to be a non-word character. `\w` covers ideographs, so this holds for the Chinese content the original bug was reported on, where the question is normally preceded by a colon or full stop.

## Tests

- `test_a_short_prompt_inside_a_longer_word_is_not_a_repetition` — Devin's example verbatim. Verified to fail against the implementation currently on `main`.
- `test_a_question_the_narration_ended_on_is_a_repetition_in_english_too` — a guard for the whitespace change, so reducing rather than removing spaces does not stop a genuine repetition from matching.

`ruff` whole-tree and the agent suite (64 passed) pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/ai-shifu/ai-shifu/pull/2919" target="_blank"><picture><source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=4">Devin Review</picture></a>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved duplicate-question detection by normalizing spacing while preserving support for scripts without spaces.
  - Prevented prompts embedded within larger words from being incorrectly treated as repeated.
  - Continued suppressing repeated prompts at the end of narration, including prompts without preceding spaces or punctuation.
- **Tests**
  - Added coverage for repeated prompts, word-boundary handling, unspaced scripts, and echo-window edge cases.

### CODE REVIEW

File: src/api/flaskr/service/learn/agent/run_agent.py [hunk 1]
```diff
@@ -388,8 +388,33 @@ def _paged(
...
+
+    Matching characters is not the same as matching what was said: a prompt reading `rate` is
+    contained in `separate`, and a lesson that had only mentioned separating examples would lose
+    the question it meant to ask, leaving the controls with nothing above them.
+    """
+    for match in re.finditer(re.escape(asked), window):
+        before = window[match.start() - 1] if match.start() else ""
+        after = window[match.end()] if match.end() < len(window) else ""
+        if not _is_word_char(before) and not _is_word_char(after):
+            return True
```
[the remark this answers is not in the document]
Comment (aichy126) [reply src/api/flaskr/service/learn/agent/run_agent.py]: Accepted, fixed in `393d2973a`.

This was the more serious of the two: the courses this ran on are written in Chinese, so the boundary check was refusing the repetition in nearly every real case. Your example is the shape the content actually takes — an introductory phrase running straight into the question, no colon between them. The one instance I had verified in the browser happened to use a colon, which is why it passed.

`_joins_a_word` now treats characters from scripts that do not separate words with spaces (CJK ideographs, kana, hangul) as non-word for this purpose, so no boundary is demanded where the script provides none. Latin text keeps the protection, so `rate` inside `separate` is still rejected.

`test_a_chinese_question_needs_no_space_before_it_to_count_as_repeated` covers it, verified to fail against the previous implementation.

[the remark this answers is not in the document]
Comment (aichy126) [reply src/api/flaskr/service/learn/agent/run_agent.py]: Accepted, fixed in `393d2973a`.

Correct: the window was a slice, so its first position always looked like a boundary, including when the slice had cut a word in half. `_stands_alone` now takes the whole narration plus the index to start searching from, and reads the characters around a match from the full string, so a window opening inside `separate` still sees the `a` before `rate`.

`test_a_window_opening_inside_a_word_does_not_invent_a_boundary` covers it. The narration is sized from `_ECHO_WINDOW_CHARS` so the window opens exactly on the `r`, which is the only place the bug shows; my first attempt at this test padded to a round number instead and passed against the broken version, so the sizing is derived rather than hard-coded.
